### DISCUSSION TITLE
Automate updating GitHub Action runners (`datasource/github-runners`)

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#27745] [Suggest an Idea]
### Tell us more.

Can we _automate_ updating the source of truth for the GitHub Action runners?

## Current situation

Right now, our source of truth is the [`actions/runner-images` repository's `README.md` file](https://github.com/actions/runner-images/blob/main/README.md), that has a table of all supported versions (beta/stable).

Renovate can't read/understand the Markdown file, and Markdown formatting is easy to break by accident.

Here are our [Renovate docs to add/promote/drop a GitHub runner](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/github-runners/readme.md).

## Upstream creates new source of truth JSON file

Would it help if we had a `runner-images.json` file on the upstream repository? I'm thinking of something like the [`node-js-schedule.json` file](https://github.com/renovatebot/renovate/blob/main/data/node-js-schedule.json) that we update with Renovate. Here's a update PR from Renovate:

- #25487

**Edit:** I'm asking the upstream maintainers if they want to create a source of truth file for us here:

- https://github.com/actions/runner-images/discussions/9484

### Work for upstream maintainers

The upstream maintainers would need to:

- decide on a JSON format
- update the JSON file when they're releasing a new _beta_ version
- update the JSON file when they're marking a version as _stable_
- update the JSON file when they're _removing_ a version

### Work for us

- Read the JSON format
- Make Renovate configuration to update the JSON file on our side

## Related issues/PRs:

- #27228
- #27292

## What do you think?

What do you maintainers think?

**Comment by rarkins** [comment]
Can it be part of a custom datasource? Not sure if we quite support enough right now but ideally this type of thing can be programmatic without needing to maintain a new datasource and/or repo

**Reply by HonkingGoose** [reply]
> Can it be part of a custom datasource? 

I'm not sure what you mean with _custom_ datasource, but we already have our _own_ datasource. See this directory: [`/lib/modules/datasource/github-runners`](https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource/github-runners).

**Reply by rarkins** [reply]
I mean dropping this datasource, or rewriting it so that it can be dynamically generated using https://docs.renovatebot.com/modules/datasource/custom/

**Reply by HonkingGoose** [reply]
Thank you for the explanation! 😄 

---

**Edit:** even if we drop the datasource, or rewrite it, we probably still need a JSON file upstream as source of truth?

**Reply by rarkins** [reply]
In theory we could fetch and parse markdown into results as long as it's structured. If it's unstructured then the only way is manual maintenance of it

**Reply by HonkingGoose** [reply]
Hmm, here's the [edit history for the upstream. Markdown file](https://github.com/actions/runner-images/commits/main/README.md). I see the last two commits alone made big changes to the ordering and naming of the versions. I suppose that's too unstructured for Renovate to rely on.

**Reply by rarkins** [reply]
Could be hard to automate then?

**Reply by HonkingGoose** [reply]
I think automating around the upstream Markdown table is going to be hard.

**Comment by viceice** [comment]
if they have such a file, then I would simply download it on the fly at runtime, like we do for the azure workflow stuff.

**Reply by HonkingGoose** [reply]
They don't have a JSON file like that Node.js version JSON file. Should I ask them to make such a file? It would make our life as maintainers easier, vs more work for upstream maintainers...

**Reply by viceice** [reply]
sure 🤗

**Reply by HonkingGoose** [reply]
I'm asking the upstream maintainers about creating a JSON file that we can use here:

- https://github.com/actions/runner-images/discussions/9484

Let's see what they think. 😄
